Add tests for GlusterFS dynamic provisioner.

humblec opened the pull request:
This bring up a heketi server pod and the server will be running
    in mock mode, the PVC creation should work, however the volume
    attachment to a pod and read/write is not part of this test. Due
    to the same reason the tests are marked as [fast].


Signed-off-by: Humble Chirammal <hchiramm@redhat.com>

<!--  Thanks for sending a pull request!  Here are some tips for you:
1. If this is your first time, read our contributor guidelines https://git.k8s.io/community/contributors/devel/pull-requests.md#the-pr-submit-process and developer guide https://git.k8s.io/community/contributors/devel/development.md#development-guide
2. If you want *faster* PR reviews, read how: https://git.k8s.io/community/contributors/devel/pull-requests.md#best-practices-for-faster-reviews
3. Follow the instructions for writing a release note: https://git.k8s.io/community/contributors/devel/pull-requests.md#write-release-notes-if-needed
4. If the PR is unfinished, see how to mark it: https://github.com/kubernetes/community/blob/master/contributors/devel/pull-requests.md#marking-unfinished-pull-requests
-->

**What this PR does / why we need it**:

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
Fixes #

**Special notes for your reviewer**:

**Release note**:
<!--  Write your release note:
1. Enter your extended release note in the below block. If the PR requires additional action from users switching to the new release, include the string "action required".
2. If no release note is required, just write "NONE".
-->
```release-note

```

humblec commented:
/release-note-none

humblec commented:
/assign @rootfs @jsafrane

rootfs commented on test/e2e/storage/volume_provisioning.go:
```diff
@@ -903,6 +1008,55 @@ func newBetaStorageClass(t storageClassTest, suffix string) *storagebeta.Storage
+			Containers: []v1.Container{
+				{
+					Name:  "glusterdynamic-provisioner",
+					Image: "quay.io/humblec/glusterdynamic-provisioner:v1.0",
```
make this in gcr.io

humblec replied:
Sure, is there a common repo or I can host it under my account ?

jsafrane replied:
Look at `test/images/*`, Kubernetes maintains Dockerfiles for images used in testing there. Someone from Google then needs to build and push the image to gcr.io.

rootfs commented on test/e2e/storage/volume_provisioning.go:
```diff
@@ -716,6 +790,37 @@ var _ = utils.SIGDescribe("Dynamic Provisioning", func() {
 	})
+
+	framework.KubeDescribe("GlusterDynamicProvisioner", func() {
+		It("should create and delete persistent volumes [fast]", func() {
```
is it really fast?

jeffvance replied:
Not fast if there is potential for 20m wait.

humblec replied:
ok. I will mark it as `slow` then.

rootfs commented on test/e2e/storage/volume_provisioning.go:
```diff
@@ -716,6 +790,37 @@ var _ = utils.SIGDescribe("Dynamic Provisioning", func() {
+			}
+			class := newStorageClass(test, ns, "gluster")
+			_, err := c.Storage().StorageClasses().Create(class)
+			defer c.Storage().StorageClasses().Delete(class.Name, nil)
```
double delete storage class

rootfs commented on test/e2e/storage/volume_provisioning.go:
```diff
@@ -716,6 +790,37 @@ var _ = utils.SIGDescribe("Dynamic Provisioning", func() {
+			By("creating a claim with an alpha dynamic provisioning annotation")
+			claim := newClaim(test, ns, "default")
+			defer func() {
+				c.Core().PersistentVolumeClaims(ns).Delete(claim.Name, nil)
```
double delete claim

rootfs commented:
/assign @jeffvance

jeffvance commented on test/e2e/storage/volume_provisioning.go:
```diff
@@ -163,6 +163,80 @@ func testDynamicProvisioning(t storageClassTest, client clientset.Interface, cla
 	return pv
 }
 
+// testGlusterDynamicProvisioning tests the same workflow of testDynamicProvisioning expect it wont attach
```
expect -> except

humblec replied:
thanks.

jeffvance commented on test/e2e/storage/volume_provisioning.go:
```diff
@@ -163,6 +163,80 @@ func testDynamicProvisioning(t storageClassTest, client clientset.Interface, cla
+	defer func() {
+		framework.Logf("deleting claim %q/%q", claim.Namespace, claim.Name)
+		// typically this claim has already been deleted
+		err = client.CoreV1().PersistentVolumeClaims(claim.Namespace).Delete(claim.Name, nil)
```
there is a helper [func](https://github.com/kubernetes/kubernetes/blob/master/test/e2e/framework/pv_util.go#L161) to delete a pvc and check for err IsNotFound.

humblec replied:
yes, used the mentioned helper function.

jeffvance commented on test/e2e/storage/volume_provisioning.go:
```diff
@@ -163,6 +163,80 @@ func testDynamicProvisioning(t storageClassTest, client clientset.Interface, cla
 }
 
+// testGlusterDynamicProvisioning tests the same workflow of testDynamicProvisioning expect it wont attach
+// pv to a pod volume and try write/read. This is due to a limitation in heketi which require raw devices to
```
Can you use existing `testDynamicProvisioning` but pass new parameter to control attach?

humblec replied:
I think it make things complex, or when I first tried tests for gluster dynamic provisioner https://github.com/kubernetes/kubernetes/pull/38048, the outcome of the discussion was to keep it seperately from cloud providers , so new function.

jeffvance commented on test/e2e/storage/volume_provisioning.go:
```diff
@@ -163,6 +163,80 @@ func testDynamicProvisioning(t storageClassTest, client clientset.Interface, cla
+	framework.ExpectNoError(client.CoreV1().PersistentVolumeClaims(claim.Namespace).Delete(claim.Name, nil))
+
+	// Wait for the PV to get deleted. Technically, the first few delete
+	// attempts may fail, as the volume is still attached to a node because
```
I thought the volume is not attached in this test based on top comment.

humblec replied:
Yep, volume is not attached and I removed this comment.

jeffvance commented on test/e2e/storage/volume_provisioning.go:
```diff
@@ -163,6 +163,80 @@ func testDynamicProvisioning(t storageClassTest, client clientset.Interface, cla
+	// Wait for the PV to get deleted. Technically, the first few delete
+	// attempts may fail, as the volume is still attached to a node because
+	// kubelet is slowly cleaning up the previous pod, however it should succeed
+	// in a couple of minutes. Wait 20 minutes to recover from random cloud
```
The potential for a 20m wait means the Describe for this test should include the tag `[Slow]`. I didn't check.

humblec replied:
Removed the 20 mins wait and kept it a 3min timeout, that should be good enough. because of the same reason `[fast]` tag has been kept.

jeffvance commented on test/e2e/storage/volume_provisioning.go:
```diff
@@ -716,6 +790,37 @@ var _ = utils.SIGDescribe("Dynamic Provisioning", func() {
 		})
 	})
+
+	framework.KubeDescribe("GlusterDynamicProvisioner", func() {
```
Need `[Slow]` here.  If all test provisioners (gce, gke, aws?) support native Gluster then no extra tag is needed

humblec replied:
not planning to mix this with other cloud providers and the operation should be fast enough to complete.

humblec commented:
/test pull-kubernetes-e2e-kops-aws

humblec commented:
/test pull-kubernetes-e2e-gce

humblec commented:
@jeffvance I have addressed the comments, can you please take a look at this again?

humblec commented:
@jeffvance any idea e2e tests are failing ? From the logs I can see GlusterDynamicServer pod has started perfectly.

~~~
I0418 09:08:03.867658       1 wrap.go:42] GET /api/v1/namespaces/e2e-tests-volume-provisioning-2k5zx/pods/glusterdynamic-provisioner-j7l4q: (25.978587ms) 200 [[e2e.test/v0.0.0 (linux/amd64) kubernetes/$Format] 35.184.32.95:48412]
~~~

humblec commented:
Thanks @jeffvance as discussed it looks like the image pull failed on permission from quay.io. I am moving this image to a different location, will see how it goes.

jeffvance commented on test/e2e/storage/volume_provisioning.go:
```diff
@@ -716,6 +786,37 @@ var _ = utils.SIGDescribe("Dynamic Provisioning", func() {
+
+			By("creating a claim with an alpha dynamic provisioning annotation")
+			claim := newClaim(test, ns, "default")
+			defer func() {
```
The claim will attempted to be deleted even if it was not successfully created. Prefer to see the defer Delete after the `Expect` below.

jeffvance replied:
Also the _suffix_ arg is not used by the `newClaim` func

humblec replied:
Moved the `defer()`  after `expect` and also defined `suffix` arg.

jeffvance commented on test/e2e/storage/volume_provisioning.go:
```diff
@@ -716,6 +786,37 @@ var _ = utils.SIGDescribe("Dynamic Provisioning", func() {
+				parameters:   map[string]string{"resturl": serverUrl},
+			}
+			class := newStorageClass(test, ns, "gluster")
+			_, err := c.Storage().StorageClasses().Create(class)
```
The sc is created here and again in `testDynamicProvisioning`

humblec replied:
Yeah, SC and Claim creation in testglusterdynamicprovisioning has been disabled now. I will check the test result again and wipe the commented code accordingly.

humblec commented:
/test pull-kubernetes-integration

jsafrane commented on test/e2e/storage/volume_provisioning.go:
```diff
@@ -163,6 +163,77 @@ func testDynamicProvisioning(t storageClassTest, client clientset.Interface, cla
+	Expect(err).NotTo(HaveOccurred())
+
+	By("checking the claim")
+	// Get new copy of the claim
```
All the checks below look like copied from somewhere else. Can you please make a new function and call it instead of copying the code around?

humblec replied:
@jsafrane the code part is from `testDynamicProvisioning` , once the workflow is complete I will functionalize and call it from `testGlusterDynamicProvisioning`.

humblec commented:
/test pull-kubernetes-e2e-kops-aws

humblec committed:
This bring up a heketi server pod and the server will be running
in mock mode, the PVC creation should work, however the volume
attachment to a pod and read/write is not part of this test. Due
to the same reason the tests are marked as [fast].

Signed-off-by: Humble Chirammal <hchiramm@redhat.com>

humblec commented:
/test pull-kubernetes-e2e-kops-aws

humblec commented:
/test pull-kubernetes-e2e-gce-device-plugin-gpu

humblec commented:
/test pull-kubernetes-typecheck

humblec commented:
/test pull-kubernetes-e2e-gce-device-plugin-gpu

humblec commented:
/test pull-kubernetes-kubemark-e2e-gce

humblec commented:
/rerun

jsafrane commented:
Why only AWS tests fail?

```
[sig-storage] Dynamic Provisioning [k8s.io] GlusterDynamicProvisioner should create and delete persistent volumes [fast] 1m47s
go run hack/e2e.go -v --test --test_args='--ginkgo.focus=\[sig\-storage\]\sDynamic\sProvisioning\s\[k8s\.io\]\sGlusterDynamicProvisioner\sshould\screate\sand\sdelete\spersistent\svolumes\s\[fast\]$'
test/e2e/storage/volume_provisioning.go:788
pvCapacity is not equal to expectedCapacity
Expected
    <int64>: 2147483648
to equal
    <int64>: 2000000000
test/e2e/storage/volume_provisioning.go:105
	
```

Why it provides the right volume size on GCE?

humblec commented:
@jsafrane this is the mystery I am trying to solve since friday including weekend :).. I tried various claim sizes without luck. The size calculation goes different in different iterations. 

For e.g# 
```
1)
	claimSize:    "2Gi",
	expectedSize: "2Gi",

test/e2e/storage/volume_provisioning.go:788
pvCapacity is not equal to expectedCapacity
Expected
    <int64>: 3000000000
to equal
    <int64>: 2147483648
test/e2e/storage/volume_provisioning.go:105

2) 

1.5Gi
2Gi

pvCapacity is not equal to expectedCapacity
Expected
    <int64>: 2000000000
to equal
    <int64>: 2147483648

3)
	claimSize:    "1.5Gi",
	expectedSize: "2G",


test/e2e/storage/volume_provisioning.go:788
pvCapacity is not equal to expectedCapacity
Expected
    <int64>: 2147483648
to equal
    <int64>: 2000000000
test/e2e/storage/volume_provisioning.go:105
```

jeffvance commented:
@humblec have you tried `aws s3 ls <object-store>` to see the pv size on aws?

humblec commented:
/rerun

humblec commented:
/test pull-kubernetes-e2e-gce

humblec commented:
/test pull-kubernetes-integration

jeffvance commented:
/test pull-kubernetes-e2e-gce
/test pull-kubernetes-bazel-test
/test pull-kubernetes-e2e-kops-aws

humblec commented:
/retest

humblec commented:
I think I got the issue, :) 
Here is my thought:  the provisioner rounduptoGiB (`	sz := int(volutil.RoundUpToGiB(capacity)))`  value and sent the request to heketi as heketi always calculate the size in `GiB`. For AWS , the value calculation in `GiB` works, however for GCE the value calculation happens in `GB`. This cause the GCE tests to fail if the requested size is in `GiB` .  I also see below source documention  in E2E:
```
			// gce or gke
			if getDefaultPluginName() == "kubernetes.io/gce-pd" {
				// using GB not GiB as e2e test unit since gce-pd returns GB,
				// or expectedSize may be greater than claimSize.
				test.claimSize = "2G"
				test.expectedSize = "2G"
			}
```

I have disabled the GCE size check in last patch set.

humblec commented:
@jsafrane @jeffvance finally it passed the tests! so it looks like the GCE size calculation assumption in previous comment is in right direction :). Please share your thought. Ideally this is good to go!

humblec commented:
Can I request your review ? :) @jeffvance @jsafrane @rootfs

jeffvance commented on test/e2e/storage/volume_provisioning.go:
```diff
@@ -98,10 +99,14 @@ func testDynamicProvisioning(t storageClassTest, client clientset.Interface, cla
-	expectedCapacity := resource.MustParse(t.expectedSize)
-	pvCapacity := pv.Spec.Capacity[v1.ResourceName(v1.ResourceStorage)]
-	Expect(pvCapacity.Value()).To(Equal(expectedCapacity.Value()), "pvCapacity is not equal to expectedCapacity")
+	if class.Provisioner == "kubernetes.io/glusterfs" && framework.ProviderIs("gke") || framework.ProviderIs("gce") {
```
`framework.ProviderIs` accepts multiple strings as parameters, so this can be written as:
```
if class.Provisioner == "kubernetes.io/glusterfs" && framework.ProviderIs("gke", "gce") {
```

humblec commented:
/test pull-kubernetes-integration
